src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/CacheHealthDecider.java

```diff
@@ -130,32 +149,80 @@ private Action computeBestAction(final NodeKey esNode, final ResourceEnum cacheT
     private Action getAction(final String actionName,
                              final NodeKey esNode,
                              final ResourceEnum cacheType,
-                             final long currentCapacityInBytes,
+                             final Long currentMaxSizeInBytes,
+                             final Long heapMaxSizeInBytes,
                              final boolean increase) {
-        if (ModifyCacheCapacityAction.NAME.equals(actionName)) {
-            return configureCacheCapacity(esNode, cacheType, currentCapacityInBytes, increase);
+        if (currentMaxSizeInBytes == null || heapMaxSizeInBytes == null) {
+            return null;
+        }
+        if (ModifyCacheMaxSizeAction.NAME.equals(actionName)) {
+            return configureCacheMaxSize(esNode, cacheType, currentMaxSizeInBytes, heapMaxSizeInBytes, increase);
         }
         return null;
     }
 
-    private ModifyCacheCapacityAction configureCacheCapacity(
+    private ModifyCacheMaxSizeAction configureCacheMaxSize(
             final NodeKey esNode,
             final ResourceEnum cacheType,
-            final long currentCapacityInBytes,
+            final long currentMaxSizeInBytes,
+            final long heapMaxSizeInBytes,
             final boolean increase) {
-        final ModifyCacheCapacityAction action =
-                new ModifyCacheCapacityAction(esNode, cacheType, currentCapacityInBytes, increase);
+        final ModifyCacheMaxSizeAction action =
+                new ModifyCacheMaxSizeAction(
+                        esNode,
+                        cacheType,
+                        currentMaxSizeInBytes,
+                        heapMaxSizeInBytes,
+                        getFieldDataCacheUpperBound(),
+                        getShardRequestCacheUpperBound(),
+                        increase);
         if (action.isActionable()) {
             return action;
         }
         return null;
     }
 
-    private long getNodeCacheCapacityInBytes(final NodeKey esNode, final ResourceEnum cacheType) {
-        // TODO: use NodeConfigurationRca to return capacity, for now returning random value in Bytes
-        if (cacheType.equals(ResourceEnum.FIELD_DATA_CACHE)) {
-            return 1000L;
+    private Long getNodeCacheMaxSizeInBytes(final NodeKey esNode, final ResourceEnum cacheType) {
+        try {
+            if (cacheType.equals(ResourceEnum.FIELD_DATA_CACHE)) {
+                return (long) getAppContext().getNodeConfigCache().get(esNode, ResourceUtil.FIELD_DATA_CACHE_MAX_SIZE);
+            }
+            return (long) getAppContext().getNodeConfigCache().get(esNode, ResourceUtil.SHARD_REQUEST_CACHE_MAX_SIZE);
+        } catch (final Exception e) {
```

Comment:
Let's restrict this only to `IlegalArgumentException` .

Reply:
Modified